src/main/groovy/cruise/umple/tasks/UmpleGenerateTask.groovy

```diff
@@ -4,20 +4,40 @@ import cruise.umple.UmpleConsoleConfig
 import cruise.umple.UmpleConsoleMain
 import cruise.umple.UmpleLanguage
 import cruise.umple.internal.tasks.DefaultUmpleOptions
-import org.gradle.api.tasks.Input
-import org.gradle.api.tasks.SourceTask
-import org.gradle.api.tasks.TaskAction
+import org.gradle.api.tasks.*
 import org.gradle.api.tasks.incremental.IncrementalTaskInputs
 
 import static com.google.common.base.Preconditions.checkNotNull
 /**
  * Created by kevin on 15/03/2017.
  */
 class UmpleGenerateTask extends SourceTask {
```

Comment:
@Nava2 The incremental build stuff appears to be working now. I think this branch can be merged into ss-config-n2, now?

Reply:
I don't think that this is properly implemented. I think Gradle needs to know _all_ files that are generated. We either need to keep track of the files created or figure out a way to have Umple tell us what files it will change/create. The intention of the incremental updates is that gradle can choose whether or not to generate new files and also what files to delete.

Reply:
> I think Gradle needs to know all files that are generated
Isn't that what the `outputDirs` member variable does?

Reply:
As far as I can tell, other than the "nice to have" refactoring of `integrationTest`, this is the only real issue that needs to be resolved before merging this branch. What do you think of my earlier comment? I agree it looks a bit funky, but by setting up the `outputDirs` member variable like that, doesn't Gradle know where all the generated files will be?

Reply:
No because it needs to know the files directly. Thats what the incremental input files parameter does.

Reply:
Looking at the example here: https://docs.gradle.org/3.1/javadoc/org/gradle/api/tasks/incremental/IncrementalTaskInputs.html

What if we have an `inputDirs` member variable that contains the locations of each of the `master` files in addition to the `outputDirs` member variable that we have now? Then we'd basically just copy the `if (!inputs.incremental)` and `inputs.outOfDate...` stuff from the example... and hopefully we'd be on our way.

Reply:
Yes, somewhat. That assumes that we want to always change all files. I don't think we have another choice. The way the umple compiler works requires us to regenerate the entire tree no matter what file changes.

So yeah, that should be the best we can do, but not the ideal case. 🙂

Reply:
Getting the incremental build stuff working for the integration tests is proving tricky. It works after I rebuild the plugin, but if I don't rebuild it the generation phase gets skipped while the verification step runs anyway (!!?!?!?!), which causes the test to fail. Even better is that the printing to standard out stuff seems to largely be broken for the integration tests, so I'm more or less doing everything blind. It might have something to do with the fact that we're using the[ `test-sets` plugin](https://github.com/unbroken-dome/gradle-testsets-plugin) for `integrationTest`, but I'm really not sure. It looks like I might be able to use `GenerationTests.groovy` as a template if I can figure how to hook everything together.

On an unrelated note, I'll be starting my final semester on Tuesday. I'm not sure how much time I'll have to contribute to this project over the semester, but I'll do my best to pop in every now and again.

The only thing I can think is to try and convert `integrationTest` into a "normal" test (using Project Runner). I'm not sure how to go about doing that, though. How are the current tests (CheckGlobals.groovy etc) hooked into the test task?